### PR TITLE
Colour fill plots fail on workspaces of monitors

### LINKED ISSUE #29035: Colour Fill Plot fails to plot grouped workspaces
# Expected behavior
- The colour fill plotting should work correctly

### Actual behavior
- The following is printed and no plot appears:
```
Traceback (most recent call last):
  File "/home/david/mantidpr/qt/applications/workbench/workbench/plugins/workspacewidget.py", line 132, in _do_plot_colorfill
    pcolormesh(names)
  File "/home/david/mantidpr/Framework/PythonInterface/mantid/plots/plotfunctions.py", line 56, in inner_func
    return func(workspaces, *args, **kwargs)
  File "/home/david/mantidpr/qt/python/mantidqt/plotting/functions.py", line 199, in pcolormesh
    pcm = pcolormesh_on_axis(ax, ws)
  File "/home/david/mantidpr/qt/python/mantidqt/plotting/functions.py", line 248, in pcolormesh_on_axis
    if use_imshow(ws):
  File "/home/david/mantidpr/qt/python/mantidqt/plotting/functions.py", line 167, in use_imshow
    difference = np.diff(y)
  File "/usr/lib/python3/dist-packages/numpy/lib/function_base.py", line 1931, in diff
    return a[slice1]-a[slice2]
TypeError: ufunc 'subtract' did not contain a loop with signature matching types dtype('<U1') dtype('<U1') dtype('<U1')
```

### Steps to reproduce the behavior

The fastest way to create the grouped workspace is:
- Download the ISIS Training data
- Open the ISIS SANS interface
- Set the user file to `training_data/loqdemo/maskfile.txt`
- Set the batch file to `training_data/loqdemo/batch_reduce.csv`
- Hit process all
- Attempt to colour fill plot and of the group workspaces

### Platforms affected
RHEL7 / Ubuntu / All?

### PR DESCRIPTION
**Pull request by Harrietbrown**
**Description of work.**

This PR fixes an issue that would cause an error is a colourfill plot was attempted on a workspace that contained only monitor spectra. This is an issue if someone attempts to do a colourfill plot on a group worspace containing one of these monitor only workspaces as it prevents the plot from being performed.

**To test:**

1. Download the ISIS Training data
2. Open the ISIS SANS interface
3. Set the user file to training_data/loqdemo/maskfile.txt
4. Set the batch file to training_data/loqdemo/batch_reduce.csv
5. Hit process all
6. Attempt to colour fill plot and of the group workspaces
7. all group workspaces should be able to be plotted.

Fixes #29035

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/PythonInterface/mantid/plots/datafunctions.py [hunk 1]
```diff
@@ -480,6 +480,13 @@ def get_matrix_2d_ragged(workspace, normalize_by_bin_width, histogram2D=False, t
             max_value = max(max_value, xtmp.max())
             diff = xtmp[1:] - xtmp[:-1]
             delta = min(delta, diff.min())
+    xtmp = workspace.readX(0)
+    if delta == np.finfo(np.float64).max:
+        delta = (xtmp[1:] - xtmp[:-1]).min()
```
Comment (martyngigg) [inline Framework/PythonInterface/mantid/plots/datafunctions.py:485]: Can we use `numpy.diff().min()` here (and on line 481)? I think it signals the intent of finding the difference a little clearer then the array index trick.

File: Framework/PythonInterface/mantid/plots/datafunctions.py [hunk 2]
```diff
@@ -584,6 +591,8 @@ def get_uneven_data(workspace, distribution):
     y = []
     nhist = workspace.getNumberHistograms()
     yvals = workspace.getAxis(1).extractValues()
+    if yvals == ['']:
```
Comment (martyngigg) [inline Framework/PythonInterface/mantid/plots/datafunctions.py:594]: Can I ask what this is protecting against? The empty string is throwing me a little.

**Review by martyngigg — CHANGES_REQUESTED** [review]
This works well. The monitor-only workspaces now don't cause an error when plotted. I loaded the file `Testing/Data/SystemTest/SANS2D/SANS2D00034505.nxs` as an additional check as it's a raw data file that only contains monitors too.

Could we add some unit tests for this behaviour?

File: Framework/PythonInterface/mantid/plots/datafunctions.py (hunk 2 above)
Comment (Harrietbrown) [reply Framework/PythonInterface/mantid/plots/datafunctions.py:594]: If a workspace does not have spectra ID then `extractValues()` returns [''], which causes issues later

Comment (martyngigg) [reply Framework/PythonInterface/mantid/plots/datafunctions.py:594]: I don't quite follow. Do you mean not having a `spectraAxis`?. `extractValues` should just pull out what values are on the vertical axis. Do you have an example that causes this?

Comment (Harrietbrown) [reply Framework/PythonInterface/mantid/plots/datafunctions.py:594]: the output workspace from the origional issue has no spectra id for its spectra

image

the first workspace here returns [''] from extractValues()

Comment (martyngigg) [reply Framework/PythonInterface/mantid/plots/datafunctions.py:594]: Thanks for the example I understand the issue now. The vertical axis is a `TextAxis` that happens to have a single empty string as its value.
I think here we are dealing with the specific case of a single empty string entry but I can create an TextAxis and I get an error plotting that too so I think we need to expand and better deal with text axes in general.

Using `CreateWorkspace` we can set a text axis with some values:

```python
CreateWorkspace(OutputWorkspace='textaxis', DataX='1,2,3,1,2,3', DataY='2,2,2,2', NSpec=2, VerticalAxisUnit='Text', VerticalAxisValues='a,b')
```

Trying a colorfill on this produces an error:

```
Traceback (most recent call last):
  File "/home/mgigg/Code/mantidproject/mantid-nightly/qt/applications/workbench/workbench/plugins/workspacewidget.py", line 132, in _do_plot_colorfill
    pcolormesh(names)
  File "/home/mgigg/Code/mantidproject/mantid-nightly/Framework/PythonInterface/mantid/plots/plotfunctions.py", line 56, in inner_func
    return func(workspaces, *args, **kwargs)
  File "/home/mgigg/Code/mantidproject/mantid-nightly/qt/python/mantidqt/plotting/functions.py", line 201, in pcolormesh
    pcm = pcolormesh_on_axis(ax, ws)
  File "/home/mgigg/Code/mantidproject/mantid-nightly/qt/python/mantidqt/plotting/functions.py", line 250, in pcolormesh_on_axis
    if use_imshow(ws):
  File "/home/mgigg/Code/mantidproject/mantid-nightly/qt/python/mantidqt/plotting/functions.py", line 169, in use_imshow
    difference = np.diff(y)
  File "/usr/lib/python3/dist-packages/numpy/lib/function_base.py", line 1931, in diff
    return a[slice1]-a[slice2]
TypeError: ufunc 'subtract' did not contain a loop with signature matching types dtype('<U1') dtype('<U1') dtype('<U1')
```

I guess we need to deal with `TextAxis` differently. [`getAxis(1).isTextAxis()`](https://docs.mantidproject.org/nightly/api/python/mantid/api/TextAxis.html#mantid.api.TextAxis.isText) will return True if that axis is text. Maybe in that case we just use the indices [0, nhist) as the Y values?

**Review by martyngigg — APPROVED** [review]
Thanks for making those additional fixes. This looks good now.

**Pull request merged**